source/History.js

```diff
@@ -0,0 +1,389 @@
+				var fn = this._handler;
+				if (typeof fn == 'function') {
+					fn = fn.bind(this._currentObj);
+				}
+				fn();
+			}
+		},
+
+		/**
+		* @private
+		*/
+		_popBackHistory: function() {
+			var bStack = this._backHistoryStack;
+			bStack.pop();
+			if (bStack.length) {
+				this._currentObj = bStack[bStack.length - 1].currentObj;
+				this._handler = bStack[bStack.length - 1].handler;
+			} else {
+				this._currentObj = this._handler = null;
+			}
+
+			this.isBackInProgress = false;
+		},
+
+		/**
+		* Request of pushstate() is pushed into queue when popstate in handling.
+		* After handling is over, we can dequeue _pushBackQueue
+		*
+		* @privae
+		*/
+		_dePushBackQueue: function() {
+			var queue = this._pushBackQueue,
+				length = queue.length,
+				item;
+			for (var i = 0; i < length; i++) {
+				item = queue.pop();
+				this._pushBackHistory(item.currentObj, item.handler);
+			}
+		},
+
+		/**
+		* There are 3 kinds of popstate event trigger.
+		* history.back(), forward() and go().
+		* However, moon.History only deal with history.back().
+		* If popstate event is triggered from other ways, we should ignore them.
+		*
+		* @public
+		*/
+		popStateHandler: function() {
+			this._isPopStateInProgress = false;
+			// Todo: We cannot prevent popstate event triggerd from history.go() or history.forward()
+			// If user call those event directly, moonstone controls may have unexpected behavior.
+			var state = !this._currentObj ? "empty"
+						: (this._ignorePopState || scope.ignoreFirstPopupEvent) ? "silence"
+						: !this._currentObj.getShowing() ? "invisible"
+						: "active";
+
+			switch (state) {
+			case 'empty':
+				break;
+			case 'silence':
+			//Popstate event should be ignored on following 2 conditions.
+			//1. When App is loaded, onpopstate event fired with null state.
+			//2. history.go(-1) triggers onpopstate event but it should be ignored.
+				scope.ignoreFirstPopupEvent = false;
+				this._ignorePopState = false;
+				this._popBackHistory();
+				break;
+			case 'invisible':
+			//Current back key target is on history and have handler too.
+			//However it is invisible.
+			//At this point, we should skip calling back key hanlder and restore history.
+				history.pushState({currentObjId: this._backHistoryStack[this._backHistoryStack.length - 1].currentObj.id}, '', '');
+				break;
+			case 'active':
+				this._callBackKeyHandler();
+				this._popBackHistory();
+				break;
+			}
+
+			this._dePushBackQueue();
+			return;
+		},
+
+		/**
+		* @private
+		*/
+		remoteBackKeyHandler: function (inSender, inEvent) {
+			if (inEvent.keySymbol == 'back' && this._currentObj && this._currentObj.getShowing()) {
+				this._callBackKeyHandler();
+				this.ignorePopState();
+			}
+			return true;
+		},
+
+		/**
+		* Abstract back key handler function.
+		* Each control which mixin moon.History should implement this method.
+		*
+		* @private
+		*/
+		backKeyHandler: function () {
```

Comment:
Should this be moved into the mixin, instead of the singleton, following a form similar to `pushBackHistory`?